File: testing/nim/var_ptr_var.nim
import strformat
import std/decls

type
  Obj = object
    f1: float
    f3: char

  CbPayload = object
    p: ptr[Obj]

proc `$`[T](p: ptr T): string = $cast[int](p)
proc `$`(p: pointer): string = $cast[int](p)

proc callback(payload: CbPayload): void =
  var obj {.byaddr.}: Obj = payload.p[]
  # payload.p[].f3 = '5'
  echo &"{obj} from {payload.p}"
  obj.f3 = '2'
  echo &"{obj} set @ {addr obj}"

var obj: Obj
echo &"{obj} @ {addr obj}"
callback(CbPayload(p: addr obj))
echo &"{obj} @ {addr obj}"

#=================================  eee  =================================#

proc getCb(data: pointer, cb: proc(clientdata: pointer) {.cdecl.}) {.cdecl.} =
  echo "executing ... "
  cb(data)

var variable: string = "---"
var intvar: int = 12

var actualData = (intv: addr intvar, strv: addr variable)
type Data = typeof(actualData)

getCb(
  addr actualdata,
  proc(clientData: pointer) {.cdecl.} =
    let data = cast[ptr[Data]](clientData)
    var strv {.byaddr.}: string = data[].strv[]
    var intv {.byaddr.}: int = data[].intv[]
    strv = "hhhh"
    intv = 124314
)

echo "strv: ", variable
echo "intv: ", intvar

File: testing/nim/wip1/var_ptr_var.cpp
#include <iostream>

struct Obj {
    float f1;
    char  f3;
};

struct CbPayload {
    Obj* p;
};

void callback(CbPayload payload) {
    Obj& obj = *(payload.p);
    obj.f3   = '1';
}

int main() {
    Obj obj;
    std::cout << obj.f3 << "\n";
    CbPayload payload;
    payload.p = &obj;
    callback(payload);
    std::cout << obj.f3 << "\n";
}
